src/ILLink.RoslynAnalyzer/DataFlow/DynamicallyAccessedMembersBinder.cs

```diff
@@ -0,0 +1,438 @@
+						continue;
+
+					if ((bindingFlags & (BindingFlags.Public | BindingFlags.NonPublic)) == BindingFlags.NonPublic && method.DeclaredAccessibility == Accessibility.Public)
+						continue;
+
+					yield return method;
+				}
+
+				if ((bindingFlags & BindingFlags.DeclaredOnly) == BindingFlags.DeclaredOnly)
+					yield break;
+
+				type = type.BaseType;
+				onBaseType = true;
+			}
+		}
+
+		public static IEnumerable<IFieldSymbol> GetFieldsOnTypeHierarchy (this ITypeSymbol thisType, OperationAnalysisContext context, Func<IFieldSymbol, bool>? filter, BindingFlags? bindingFlags = BindingFlags.Default)
+		{
+			ITypeSymbol? type = thisType;
+			bool onBaseType = false;
+			while (type != null) {
+				foreach (var member in type.GetMembers ()) {
+					if (!(member is IFieldSymbol field))
+						continue;
+					// Ignore private fields on a base type - those are completely ignored by reflection
+					// (anything private on the base type is not visible via the derived type)
+					if (onBaseType && field.DeclaredAccessibility == Accessibility.Private)
+						continue;
+
+					// Note that compiler generated fields backing some properties and events will get through here.
+					// This is intentional as reflection treats these as fields as well.
+
+					if (filter != null && !filter (field))
+						continue;
+
+					if ((bindingFlags & (BindingFlags.Instance | BindingFlags.Static)) == BindingFlags.Static && !field.IsStatic)
+						continue;
+
+					if ((bindingFlags & (BindingFlags.Instance | BindingFlags.Static)) == BindingFlags.Instance && field.IsStatic)
+						continue;
+
+					if ((bindingFlags & (BindingFlags.Public | BindingFlags.NonPublic)) == BindingFlags.Public && field.DeclaredAccessibility != Accessibility.Public)
+						continue;
+
+					if ((bindingFlags & (BindingFlags.Public | BindingFlags.NonPublic)) == BindingFlags.NonPublic && field.DeclaredAccessibility == Accessibility.Public)
+						continue;
+
+					yield return field;
+				}
+
+				if ((bindingFlags & BindingFlags.DeclaredOnly) == BindingFlags.DeclaredOnly)
+					yield break;
+
+				type = type.BaseType;
+				onBaseType = true;
+			}
+		}
+
+		public static IEnumerable<ITypeSymbol> GetNestedTypesOnType (this ITypeSymbol type, Func<ITypeSymbol, bool>? filter, BindingFlags? bindingFlags = BindingFlags.Default)
+		{
+			// @TODO
+			throw new NotImplementedException ("@TODO");
+			//type.InheritsFromOrEquals()
+			//foreach (var nestedType in type.NestedTypes) {
+			//	ITypeSymbol
+			//	if (filter != null && !filter (nestedType))
+			//		continue;
+
+			//	if ((bindingFlags & (BindingFlags.Public | BindingFlags.NonPublic)) == BindingFlags.Public) {
+			//		if (!nestedType.IsNestedPublic)
+			//			continue;
+			//	}
+
+			//	if ((bindingFlags & (BindingFlags.Public | BindingFlags.NonPublic)) == BindingFlags.NonPublic) {
+			//		if (nestedType.IsNestedPublic)
+			//			continue;
+			//	}
+
+			//	yield return nestedType;
```

Comment:
I think what we want is to use [GetTypeMembers](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.inamespaceortypesymbol.gettypemembers?view=roslyn-dotnet-4.0.1) which returns an enumeration of the type inside the symbol.

Reply:
Thanks!